### ISSUE TITLE
Enqueued transitions for the FSM

### ISSUE THREAD

**Issue by dastillero** [KybernetikGames/animancer#79]
# Use Case

I have an use case where a number of other states can be candidates to be the next one. I use to write the logic to choose the correct state at the end of the Update() or LateUpdate() methods. Once the correct state has been chosen, I call TrySetState() or ForceSetState().

# Solution

It would be nice to be able to enqueue several candidate states and let the state machine choose the correct one at the end of each update cycle based on the outcome of the CanEnterState() method and the priority of each candidate state.

As several of the candidate states might return true in their CanEnterState() method, some sort of priority should be applied. States may implement a GetPriority() method to determine the priority for each one, while the state machine can use a priority queue to sort the candidate states by that priority.

If several states return the same priority, the state machine can simply choose the first one in the priority queue.

# Alternatives

Instead of letting the state machine choose which of the candidate states should be the next one at the end of each update cycle, states machines could expose an EvalEnqueuedStates() method which could be called by the states whenever is necessary.

**Comment by KybernetikGames** [comment]
Animancer's state machines don't have an update cycle, they change states immediately when you tell them to and do nothing otherwise.

But what you're looking for sounds quite similar to the [Input Buffer](https://kybernetik.com.au/animancer/v6/docs/manual/fsm#input-buffers) system which is a separate class that targets a `StateMachine` and has an `Update` method for you to call wherever is appropriate (Update/Fixed/Late).

It sounds like you don't really want a "first in - first out" queue though, but rather a place to group a bunch of states which can later be told to try entering them in order of priority. If so, then `StateSelector` might be a good name for it.

Some things to consider:

- Does it clear the list when it successfully enters a state?
- Can states change priority after you add them to the list?
  - If yes, then you would need an `IPriorityState` interface for states with a `GetStatePriority` method.
  - If no, then `StateSelector.Add` could just take separate parameters for the priority and the state. You could also have the same interface with a one parameter `Add` method, but the comments would need to explain that changing the value afterwards won't affect the selector.

**Comment by dastillero** [comment]
In my case, I have extended StateMachine to have an update cycle. I have a
component attached to my character which calls an Update() method in my
custom state machine (which is implementing IUpdatable). I am still using
TrySetState() when I want to change states immediately. But all this is
specific to my use case. I think that FSM should remain generic and do not
depend on an update cycle.

I never thought of the InputBuffer class to implement something like this.
I am using InputBuffers to chain combos and things like that but I can try
them for this use case.

And yes, it is not a "first in-first out" queue. I called priority queue
because in Java there is a collection called that way which does exactly
what I wanted. In my implementation I do not actually use a priority queue
implementation. I am evaluating all the candidate states which return true
in their CanEnterState() method and then I sort them based on the outcome
of a GetPriority() method. After that, I am choosing the first one in the
sorted collection. Every state implements an IPriorizable interface
(equivalent to the IPriorityState you mention). Priorities are dynamic in
some way. For instance, I am using this same system to randomly choose
between alternative idle animations. To do this, the GetPriority() method
of each one of the candidate states returns a random number. But this is
something which you could also achieve by setting a priority in the
StateSelector.Add() method.

In my case, the Queue is being cleared on every update cycle (as I am
implementing an update cycle). But clearing it when a state change happens
should be enough.

In summary, the approach you are describing with the StateSelector,
IPriorityState and so on is correct. I do not have a case where priorities
change after being added to the StateSelector, so a StateSelector.Add()
method could also do the job.

Best regards,
Daniel.

**Comment by dastillero** [comment]
Ups. Sorry I closed this by mistake. I have reopened it.

I am thinking on this affirmation I made: "In my case, the Queue is being cleared on every update cycle (as I am
implementing an update cycle). But clearing it when a state change happens
should be enough."

In my case it would probably work, but maybe the most flexible approach would be to have a method to clear the StateSelector manually if it is needed.

**Comment by KybernetikGames** [comment]
`IPriorizable` is definitely a better name for it.

It could give you something along the lines of:

```
if (selector.TryEnterState())
    selector.Clear();
```

But I'm struggling to think of a scenario where you would want to collect potential transitions in one state and then they would still be relevant after one of them is successful. And one of those potential transitions would be the state you are now in. I'm thinking it would be better to just clear it and let the caller refill it with whatever is relevant.

**Comment by dastillero** [comment]
I was thinking more on implementing both, the automatic clearing of the StateSelector when another state is entered, and the manual method to clear the StateSelector if it is needed. As you said, I cannot imagine a circumstance where the collected transitions are still relevant after a state change.

I am clearing manually the StateSelector on every update cycle because if there is no state change, the previously collected transitions remain available in the StateSelector and that was a source of bugs in my case. The logic which add the transitions to the StateSelector is executed every frame, so I decided to clear the StateSelector on every update cycle to avoid tracking which transitions were previously added to the StateSelector. It can also happen that in one frame one of the transitions is interesting to be added to the StateSelector, but the next frame is not interesting anymore. That is the reason why I am asking for the manual method to clear the StateSelector in addition to the automatic clearing after a state change happens.

**Comment by KybernetikGames** [comment]
<details><summary>Here's my first attempt at a <code>StateSelector</code> class:</summary>

```
// Animancer // https://kybernetik.com.au/animancer // Copyright 2020 Kybernetik //

using System;
using UnityEngine;

namespace Animancer.FSM
{
    public interface IPrioritizable : IState
    {
        public float Priority { get; }
    }

    public partial class StateMachine<TState>
    {
        /// <summary>A prioritised list of potential states for a <see cref="StateMachine{TState}"/> to enter.</summary>
        public class StateSelector
        {
            /************************************************************************************************************************/

            private float[] _Priorities;

            private TState[] _States;

            /************************************************************************************************************************/

            /// <summary>The <see cref="StateMachine{TState}"/> this selector is controlling.</summary>
            public StateMachine<TState> StateMachine { get; set; }

            public int Count { get; private set; }

            public int Capacity
            {
                get => _Priorities.Length;
                set
                {
                    if (Count == 0)
                    {
                        _Priorities = new float[value];
                        _States = new TState[value];
                    }
                    else
                    {
                        var newPriorities = new float[value];
                        var newStates = new TState[value];

                        var copy = Mathf.Min(Count, value);
                        Array.Copy(_Priorities, newPriorities, copy);
                        Array.Copy(_States, newStates, copy);

                        _Priorities = newPriorities;
                        _States = newStates;
                    }
                }
            }

            /************************************************************************************************************************/

            /// <summary>Creates a new <see cref="StateSelector"/>.</summary>
            public StateSelector(StateMachine<TState> stateMachine, int capacity = 8)
            {
                _Priorities = new float[capacity];
                _States = new TState[capacity];
                StateMachine = stateMachine;
            }

            /************************************************************************************************************************/

            public void Add(float priority, TState state)
            {
                if (Capacity == Count)
                {
                    var capacity = Capacity;
                    if (capacity < 16)
                        capacity = 16;

                    while (capacity <= Count)
                        capacity *= 2;

                    Capacity = capacity;
                }

                var index = 0;
                for (; index < Count; index++)
                {
                    if (priority < _Priorities[index])
                    {
                        Array.Copy(_Priorities, index, _Priorities, index + 1, Count - index);
                        Array.Copy(_States, index, _States, index + 1, Count - index);
                        break;
                    }
                }

                _Priorities[index] = priority;
                _States[index] = state;

                Count++;
            }

            public void Add<TPrioritizable>(TPrioritizable prioritizable)
                where TPrioritizable : TState, IPrioritizable
            {
                Add(prioritizable.Priority, prioritizable);
            }

            /************************************************************************************************************************/

            public bool CanSetState()
            {
                for (int i = Count - 1; i >= 0; i--)
                    if (StateMachine.CanSetState(_States[i]))
                        return true;

                return false;
            }

            /************************************************************************************************************************/

            public bool TrySetState()
            {
                for (int i = Count - 1; i >= 0; i--)
                {
                    if (StateMachine.TrySetState(_States[i]))
                    {
                        Clear();
                        return true;
                    }
                }

                return false;
            }

            /************************************************************************************************************************/

            public bool TryResetState()
            {
                for (int i = Count - 1; i >= 0; i--)
                {
                    if (StateMachine.TryResetState(_States[i]))
                    {
                        Clear();
                        return true;
                    }
                }

                return false;
            }

            /************************************************************************************************************************/

            public void ForceSetState()
            {
                StateMachine.ForceSetState(_States[Count - 1]);
                Clear();
            }

            /************************************************************************************************************************/

            public void Clear()
            {
                Array.Clear(_Priorities, 0, Count);
                Array.Clear(_States, 0, Count);
                Count = 0;
            }

            /************************************************************************************************************************/
        }
    }
}
```
</details>

**Comment by dastillero** [comment]
That was fast. I will set up a project with Animancer 6.0 alpha and this to
try them.

I am still using Unity 2019 LTS. No support for C# 8.0, so I changed:

    public interface IPrioritizable : IState
    {
        public float Priority { get; }
    }

to:

    public interface IPrioritizable : IState
    {
        float Priority { get; }
    }


Regards.

**Comment by KybernetikGames** [comment]
Yeah, I edited that right after posting. I thought it was fine because VS didn't give any errors but then Unity complained.

**Comment by dastillero** [comment]
Hi:

I have been trying it. It works well for all the use cases I am dealing with. The only suggestion I have is to make the setter of the StateMachine property private. I can not imagine any use case where you want to change the StateMachine which a StateSelector is controlling.

Regards.

**Comment by KybernetikGames** [comment]
[Animancer v6.0 Alpha 3](https://kybernetik.com.au/animancer/v6/docs/changes/animancer-v6-0/test#alpha-3) is now available and contains the new `StateSelector` class which has been cleaned up a bit since I posted it here.

**Comment by KybernetikGames** [comment]
[Animancer v6.0](https://kybernetik.com.au/animancer/docs/changes/animancer-v6-0/) is now up on [itch.io](https://kybernetik.itch.io/animancer-pro) (preferred) and the [Unity Asset Store](https://assetstore.unity.com/packages/tools/animation/animancer-pro-116514).
